lib/events.js

```diff
@@ -512,3 +513,76 @@ function once(emitter, name) {
     emitter.once(name, eventListener);
   });
 }
+
+const AsyncIteratorPrototype = Object.getPrototypeOf(
+  Object.getPrototypeOf(async function* () {}).prototype);
+
+function createIterResult(value, done) {
+  return { value, done };
+}
+
+function on(emitter, event) {
+  const unconsumedEvents = [];
+  const unconsumedPromises = [];
+  let error = null;
+
+  const iterator = Object.setPrototypeOf({
+    next() {
+      if (error) {
+        return Promise.reject(error);
+      }
+
+      const value = unconsumedEvents.shift();
+      if (value) {
+        return Promise.resolve(createIterResult(value, false));
+      }
+
+      return new Promise(function(resolve, reject) {
+        unconsumedPromises.push({ resolve, reject });
+      });
+    },
+
+    return() {
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+
+      return Promise.resolve(createIterResult(undefined, true));
+    },
+
+    throw(err) {
+      // TODO should we validate that err exists?
```

Comment:
No, I think we shouldn't since it is allowed to throw `undefined`. I really hope users don't actually do it.